search as you type fieldmapper

andyb-elastic opened the pull request:
Adds a new field type, `search_as_you_type`, that acts like a text field optimized for as-you-type search completion. It creates a couple subfields that analyze the indexed terms as shingles, against which full terms are queried, and a prefix subfield that analyze terms as the largest shingle size used and edge-ngrams, against which partial terms are queried

Adds a `boolean_prefix` query type that creates a boolean clause of a term query for each term except the last, for which a boolean clause with a prefix query is created. This will be used as the recommended query type for querying `search_as_you_type` fields, although other text queries will be supported as well

This is for #33160

andyb-elastic commented:
@jimczi from our discussion yesterday I 

* set default max_shingle_size to 3
* capped it at 5
* removed configurability for min_gram and max_gram
* added tests for the configuration options exposed

I think this is ready for a first look unless there's anything else you'd like me to add

andyb-elastic commented on modules/search-as-you-type/src/main/java/org/elasticsearch/index/mapper/SearchAsYouTypeFieldMapper.java:
```diff
@@ -0,0 +1,432 @@
+            final SuggesterizedFieldType withEdgeNgrams = new SuggesterizedFieldType(name() + "._with_edge_ngrams");
+            final SearchAsYouTypeAnalyzer wrappedWithEdgeNGrams = SearchAsYouTypeAnalyzer.withEdgeNGrams(indexAnalyzer);
+            withEdgeNgrams.setIndexAnalyzer(new NamedAnalyzer(indexAnalyzer.name(), AnalyzerScope.INDEX, wrappedWithEdgeNGrams));
+            withEdgeNgrams.setSearchAnalyzer(indexAnalyzer);
```
Now wondering if we should set this subfield's search analyzer to be a SearchAsYouTypeAnalyzer with both singles and edge ngrams turned off (e.g. a no-op analyzer wrapper), just so that we have the consistency of all SuggesterizedFieldTypes' analyzers being the same type

jimczi replied:
Are you referring to the root field ?

andyb-elastic replied:
I was referring to the subfield without shingles and with edge ngrams, but now considering the root field too

jimczi commented on server/src/main/java/org/apache/lucene/index/FreqIsScoreSimilarity.java:
```diff
@@ -0,0 +1,16 @@
+package org.apache.lucene.index;
```
Let's remove this class, it is unused

andyb-elastic replied:
++

jimczi commented on server/src/main/java/org/elasticsearch/index/mapper/MapperService.java:
```diff
@@ -292,6 +292,7 @@ public void merge(Map<String, Map<String, Object>> mappings, MergeReason reason)
                 mappingSourcesCompressed.put(entry.getKey(), new CompressedXContent(Strings.toString(
                     XContentFactory.jsonBuilder().map(entry.getValue()))));
             } catch (Exception e) {
+                logger.info("problem", e);
```
is there a problem ? ;)

jimczi commented on server/src/main/java/org/elasticsearch/index/mapper/OldSearchAsYouTypeFieldMapper.java:
```diff
@@ -0,0 +1,298 @@
+/*
```
Let's remove this one ;)

jimczi commented on modules/search-as-you-type/src/main/java/org/elasticsearch/index/mapper/SearchAsYouTypeFieldMapper.java:
```diff
@@ -0,0 +1,432 @@
+        }
+    }
+
+    public static class SuggesterizedFieldType extends TermBasedFieldType {
```
We need something to infer the type of terms that are indexed in this field (`max_shingle` and if it contains edge ngrams or not).

jimczi commented on modules/search-as-you-type/src/main/java/org/elasticsearch/index/mapper/SearchAsYouTypeFieldMapper.java:
```diff
@@ -0,0 +1,432 @@
+        }
+    }
+
+    public static class SuggesterizedFieldMapper extends FieldMapper implements ArrayValueMapperParser { // todo better name
```
Same here, we need something to get the informations of what is indexed internally.

jimczi replied:
We should also have a way to get these mappers (or field types) directly from the main mapper:
````
public SuggesterizedFieldType getSubType(int shingleSize, boolean withEdgeNGram);
````
?

andyb-elastic replied:
I added this method - I added the analysis details to the fieldtype, I think that's enough since you can do mapper.fieldType().shingleSize()

jimczi commented on modules/search-as-you-type/src/test/java/org/elasticsearch/index/mapper/SearchAsYouTypeFieldMapperTests.java:
```diff
@@ -0,0 +1,188 @@
+        }
+
+        assertThat(indexAnalyzer.isWithEdgeNGrams(), equalTo(withEdgeNGrams));
+    }
```
Can you add some tests to check the parsing and the output of the documents (make sure that all shingles/edge ngrams are correctly created) ?

andyb-elastic replied:
++

jimczi commented on modules/search-as-you-type/src/main/java/org/elasticsearch/index/mapper/SearchAsYouTypeFieldMapper.java:
```diff
@@ -0,0 +1,432 @@
+        // field type is which
+        // eg we could make a SuggesterizedFieldTypeGroup class that can look up the only edge ngreams field and then also do
+        // number of shingles, true/false edge ngrams
+        // the alternative is do a bunch of string math
```
+1 for a simple way to retrieve the shingle and edge ngram fields easily. We should avoid string math ;).

jimczi commented on modules/search-as-you-type/src/main/java/org/elasticsearch/index/mapper/SearchAsYouTypeFieldMapper.java:
```diff
@@ -0,0 +1,432 @@
+                } else if (fieldName.equals("index_options")) {
+                    builder.indexOptions(nodeIndexOptionValue(fieldNode));
+                    iterator.remove();
+                } else if (fieldName.equals("fields")) {
```
We should not handle multi fields here, we already have sub-fields so no need to add more complexity ?

andyb-elastic replied:
Strongly agree

jimczi commented on modules/search-as-you-type/src/main/java/org/elasticsearch/index/mapper/SearchAsYouTypeFieldMapper.java:
```diff
@@ -0,0 +1,432 @@
+                    maxShingleSize = XContentMapValues.nodeIntegerValue(fieldNode);
+                    iterator.remove();
+                }
+            }
```
Should we throw an error if the map is not empty (has some fields that we don't support) ?

andyb-elastic replied:
I assumed whatever calls this method did that afterwards, it looks like in at least some places they do. Should I throw one here too? https://github.com/elastic/elasticsearch/blob/3c03975ce396ff3e6f27b79d26e77a4cf1081f67/server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java#L276-L284

romseygeek commented:
I don't think we need to use all these different fields, as they're mostly duplicating tokens from each other.  A single field with shingles and edge n-grams ought to cover everything.

For example, a fixed shingle filter of three + edge ngrams when given the tokenstream "aa bb cc dd ee" will produce the following:
`a`, `aa`, `aa `, `aa b`, `aa bb`, `aa bb `, `aa bb c`, `aa bb cc`, `b`, `bb`, `bb `, `bb c`, `bb cc`, `bb cc `, `bb cc d`, `bb cc dd`, `c`, `cc`, `cc `, `cc d`, `cc dd`, `cc dd e`, `cc dd ee`
A fixed shingle filter of two + edge ngrams produces this:
`a`, `aa`, `aa `, `aa b`, `aa bb`, `b`, `bb`, `bb `, `bb c`, `bb cc`, `c`, `cc`, `cc `, `cc d`, `cc dd`, `d`, `dd`, `dd `, `dd e`, `dd ee`
- all of the latter are also produced by the former, with the exception of the final set of tokens starting with `d`.  We can trick the 3-shingle filter into also producing these ones by adding two extra positions to the end of the tokenstream.  This can be done by writing a new TokenFilter with an `incrementToken()` that just calls `in.incrementToken()`, and setting the position increment to `2` in `end()`.

andyb-elastic commented:
I see what you mean, a majority of the tokens are duplicated at least once in your example, and some are duplicated up to 4 times

Also agree from a usability and footgun-avoidance perspective, it would be surprising that adding this field type to your mapping actually creates up to ten fields. And it would be reasonable for users to want to use several of these fields in a single index.

>All of the latter are also produced by the former, with the exception of the final set of tokens starting with d. We can trick the 3-shingle filter into also producing these ones by adding two extra positions to the end of the tokenstream. This can be done by writing a new TokenFilter with an incrementToken() that just calls in.incrementToken(), and setting the position increment to 2 in end().

I'm giving this a shot. So in this case I think we end up with two fields - the root one which does not modify analysis, and the "other" one which wraps it with shingles and edge ngrams

jimczi commented:
> I'm giving this a shot. So in this case I think we end up with two fields - the root one which does not modify analysis, and the "other" one which wraps it with shingles and edge ngrams

I don't think we should pack all shingles and edge ngrams in the same field. The idea that Alan proposed is to use a single edge ngrams field that use the max shingle size. So in the end you'd have one field per shingle size (1, 2 and 3 for instance) and 1 field for the edge ngrams that use the max shingle size (3 in this example). This avoids the duplication since the edge ngrams for the 3-shingles will also produce the prefixes for the 1-shingle and 2-shingle.

andyb-elastic commented:
I pushed changes to create one subfield analyzed with shingles for each shingle size, and a single subfield that's analyzed with shingles of max shingle size and edge ngrams. This reduces the maximum amount of fields that are created for this type down to 6. There's still some duplication of tokens between the field with shingles and edge ngrams and a couple of the shingles and no edge ngram fields but it's not as bad as before

This made me think, would it make more sense to collapse all of the shingles-and-no-edge-ngrams fields into one field using the variable size ShingleFilter (rather than FixedShingleFilter like I've been doing)? I think we can configure it to produce the same tokens as we'd get from multiple fixed-size-shingle fields, and we'd only have a constant three fields created for this datatype (the root field, the variable-size shingles, and the max-shingle-size fixed-size shingles with edge ngrams)

There's some more cleanup and code quality work to be done here but I wanted to get the implementation sorted before doing that

andyb-elastic commented:
Jenkins run tests please

andyb-elastic commented:
Okay I pushed the new field scheme we talked about. Next I'll start on adding positions to the analyzer so that we get all the terms we need, and move it to core.

Also since we've decided to make these against master I've changed the base branch and deleted our feature branch

andyb-elastic commented:
Jenkins run tests please

andyb-elastic commented:
Jenkins test this please

andyb-elastic commented:
Jenkins run the docs check please

jimczi commented on modules/search-as-you-type/src/main/java/org/elasticsearch/index/mapper/SearchAsYouTypeFieldMapper.java:
```diff
@@ -0,0 +1,558 @@
+
+    public static final String CONTENT_TYPE = "search_as_you_type";
+    private static final int LOWEST_MAX_SHINGLE_SIZE = 2;
+    private static final int HIGHEST_MAX_SHINGLE_SIZE = 4;
```
ok

andyb-elastic commented:
@jimczi I pushed docs that describe the field structure

andyb-elastic commented on docs/reference/mapping/types/search-as-you-type.asciidoc:
```diff
@@ -0,0 +1,184 @@
+
+`my_field._index_prefix`::
+
+    Wraps the analyzer of `my_field._2gram` with an edge ngram token filter
```
I omitted an explanation of the filter that adds position increments at the end to create the trailing shingles - I thought it might be confusing here and better illustrated in example queries

jimczi replied:
++, I expect this field to be used only for expert queries. It should also not be very useful to document it once the `match_prefix` query is added. Maybe we should omit it but add a section regarding how prefix queries are optimized ?

andyb-elastic commented on docs/reference/mapping/types/search-as-you-type.asciidoc:
```diff
@@ -0,0 +1,184 @@
+
+
+[[general-params]]
+==== Parameters of the field type as a text field
```
I noted how we set these when setting up the field, but I wonder if putting this information here is not appropriate since we're not testing all of these settings in the field mapper tests

jimczi replied:
let's add some tests then. We should at least ensure that custom analyzers and index options work.

jimczi replied:
I think we can just add a link here (we still need tests though ;) ).

jimczi commented on docs/reference/mapping/types/search-as-you-type.asciidoc:
```diff
@@ -0,0 +1,184 @@
+`my_field._2gram`::
+
+    Wraps the analyzer of `my_field` with a shingle token filter of shingle
+    size 2
```
The default is to build shingles up to size 3 so maybe add an explanation regarding the rule to name the shingle fields ?

jimczi commented on docs/reference/mapping/types/search-as-you-type.asciidoc:
```diff
@@ -0,0 +1,184 @@
+    creating one subfield for each shingle size between 2 and
+    `max_shingle_size`. Accepts integer values between 2 (default) and 4
+    inclusive.
+
```
Can you add an example query that targets the different shingles. Something like a `match` query with an `and` operator on the root field in a `must` clause and then one `match` query per shingle size on should clauses with operator set to `or` ? This would explain how we expect to use the shingle fields (adds a cheap proximity boost).

andyb-elastic replied:
👍 I added this query

jimczi replied:
I added the support for `phrase` and `phrase_prefix` query in https://github.com/elastic/elasticsearch/pull/35600/commits/9532c19a64b685a89de017354ef599c880b48bfc so I think we can show a simple example with a `match_phrase_prefix` on the root field.

andyb-elastic commented:
@jimczi I added tests for the rest of the settings and added a few places where the prefix field gets them as well, can you take a look

andyb-elastic commented:
Jenkins run elasticsearch-ci/2 please

andyb-elastic commented:
Also note to self as part of this to update the docs page on match phrase prefix to mention this since it talks about search as you type https://github.com/elastic/elasticsearch/blame/5f106a27eaaf6deded97b421230065aa3b487409/docs/reference/query-dsl/match-phrase-prefix-query.asciidoc#L45

jimczi commented:
@andyb-elastic what is the status of this pr ? Can you update the description and share the latest updates ?

andyb-elastic commented:
@jimczi I pushed a test for building boolean prefix query, let's talk about this in more detail tomorrow

jimczi commented on docs/reference/mapping/types/search-as-you-type.asciidoc:
```diff
@@ -0,0 +1,267 @@
+--------------------------------------------------
+// CONSOLE
+// TEST[continued]
+
```
Since we've decided to not expand the search automatically to the shingle field I think we need an example with a `multi_match` that targets the root field and the shingle field with the type set to `match_phrase_prefix`, WDYT ?

jimczi commented on server/src/test/java/org/elasticsearch/index/query/BooleanPrefixQueryBuilderTests.java:
```diff
@@ -0,0 +1,226 @@
+import static org.hamcrest.Matchers.hasProperty;
+import static org.hamcrest.Matchers.hasSize;
+
+public class BooleanPrefixQueryBuilderTests extends AbstractQueryTestCase<BooleanPrefixQueryBuilder> {
```
Can you also add a test in `MultiMatchQueryBuilderTests` that checks the `boolean_prefix` type ?

andyb-elastic replied:
Added

jimczi commented on modules/mapper-extras/src/test/resources/rest-api-spec/test/search-as-you-type/20_highlighting.yml:
```diff
@@ -0,0 +1,74 @@
+  - match: { hits.total: 1 }
+  - match: { hits.hits.0._source.a_field: "quick red fox lazy brown dog" }
+  - match: { hits.hits.0._source.text_field: "quick red fox lazy brown dog" }
+  - match: { hits.hits.0.highlight.a_field.0: "quick <em>red</em> fox lazy brown dog" }
```
Can you add a test with a `multi_match` query that targets all shingle fields and sets the type to `boolean_prefix` ?

romseygeek commented on modules/mapper-extras/src/test/java/org/elasticsearch/index/mapper/SearchAsYouTypeFieldMapperTests.java:
```diff
@@ -0,0 +1,700 @@
+            expected.add(new Term("field._3gram", "two three four"));
+            //expected.add(new Term("field._3gram", "three"));
+            //expected.add(new Term("field._3gram", "four"));
+            logger.error("PREFIX ACTUAL " + actual.getClass() + " " + actual);
```
This looks like a leftover?

andyb-elastic replied:
Thanks, missed this

andyb-elastic commented on server/src/test/java/org/elasticsearch/index/query/MultiMatchQueryBuilderTests.java:
```diff
@@ -240,6 +242,51 @@ public void testToQueryFieldMissing() throws Exception {
+        assertThat(assertBooleanSubQuery(booleanQuery, PrefixQuery.class, 1).getPrefix(), equalTo(new Term(STRING_FIELD_NAME, "bar")));
+    }
+
+    public void testToQueryBooleanPrefixMultipleFields() throws IOException {
```
@jimczi I think this behavior looks correct - with a text field and an alias field, we get a dismax with two boolean_prefix queries. With a text field and a keyword field, we get a dismax with a boolean_prefix on the text field and a prefix query on the keyword field

jimczi replied:
I agree, this is the expected behavior

andyb-elastic commented on modules/mapper-extras/src/test/resources/rest-api-spec/test/search-as-you-type/20_highlighting.yml:
```diff
@@ -0,0 +1,202 @@
+  - match: { hits.hits.0._source.a_field: "quick red fox lazy brown dog" }
+  - match: { hits.hits.0._source.text_field: "quick red fox lazy brown dog" }
+  - match: { hits.hits.0.highlight.a_field: ["quick <em>red</em> <em>fox</em> <em>lazy</em> brown dog"] }
+  - match: { hits.hits.0.highlight.a_field\._2gram: ["quick <em>red fox</em><em> lazy</em> brown dog"] }
```
This highlighting behavior on the shingle fields smaller than the matching one is a little questionable

jimczi replied:
Highlighting should be done on the root field so I don't think it's an issue.

andyb-elastic commented:
Still to do with docs
* Fill out more detail for the bool prefix type in multi match
* Discussion about bool prefix and other query types in the search as you type field page
* Clarify recommendations about search as you type in the other full text query pages

andyb-elastic commented:
@jimczi I believe I've covered all the docs changes we discussed, do you mind taking a look at the changes since your last review

When going through what multi match parameters are supported for bool_prefix, it looks like the `operator`, `auto_generate_synonyms_phrase_query`, and the fuzziness family of settings are applied to query in the `multi_match` context, so it seems like we should support them in the single-query form too. I think we can leave those as a separate PR unless you'd like them included here. I explicitly disallowed the multi_match params that this query doesn't support so that they're not silently ignored.

andyb-elastic commented on server/src/main/java/org/elasticsearch/index/query/MultiMatchQueryBuilder.java:
```diff
@@ -687,6 +692,16 @@ public static MultiMatchQueryBuilder fromXContent(XContentParser parser) throws
+            throw new ParsingException(parser.getTokenLocation(),
+                "[" + CUTOFF_FREQUENCY_FIELD.getPreferredName() + "] not allowed for type [" + type.parseField.getPreferredName() + "]");
+        }
+
```
I'm not sure this is really the ideal place to be enforcing this, but I put it here for consistency with the existing restriction on fuzziness

jimczi commented on docs/reference/mapping/types/search-as-you-type.asciidoc:
```diff
@@ -0,0 +1,299 @@
+    "properties": {
+      "my_field": {
+        "type": "search_as_you_type",
+        "max_shingle_size": 4
```
Let's use the default here, 3 is enough and we should not encourage users to increase this value if they don't need to.

andyb-elastic replied:
I'll remove this example and make the following examples use the max_shingle_size = 3 index

jimczi commented on docs/reference/query-dsl/multi-match-query.asciidoc:
```diff
@@ -516,3 +519,55 @@ per-term `blended` queries. It accepts:
+[[type-bool-prefix]]
+==== `bool_prefix`
+
+The `bool_prefix` type's scoring behaves like <<type-best-fields>>, but using a
```
This is not correct, we use the `most-fields` mode for this query so we sum all field scores like a normal disjunction. The `tie_breaker` option can be used to change this behavior so I wonder if we should rename `best_fields` and `most_fields` to `bool` and rely on this option solely. That's not in the scope of this pr though so let's fix this new docs for now.

andyb-elastic replied:
Thanks for catching this, I see what my misunderstanding about the code here was

jimczi commented on docs/reference/query-dsl/multi-match-query.asciidoc:
```diff
@@ -516,3 +519,55 @@ per-term `blended` queries. It accepts:
+--------------------------------------------------
+// CONSOLE
+
+would be roughly equivalent to the query
```
There is an example for `most_fields` already so let's not repeat here if we fix the previous link ?

andyb-elastic replied:
That makes sense, I'll remove this example

jimczi commented on server/src/main/java/org/elasticsearch/index/query/BoolPrefixQueryBuilder.java:
```diff
@@ -0,0 +1,214 @@
+ * The boolean prefix query analyzes the input text and creates a boolean query containing a Term query for each term, except
+ * for the last term, which is used to create a prefix query
+ */
+public class BoolPrefixQueryBuilder extends AbstractQueryBuilder<BoolPrefixQueryBuilder> {
```
Should we rename to `MatchBoolPrefixQueryBuilder` for consistency ?

andyb-elastic replied:
I think the original user-facing name you mentioned was `match_prefix` which would be consistent with the other match queries. And the behavior would be what you'd expect based on an understanding of what the other match queries do

jimczi replied:
Agreed but I think I prefer `match_bool_prefix` which is more verbose but also more consistent with the other names.

andyb-elastic replied:
Sounds good let's go with `match_bool_prefix`

andyb-elastic commented:
>I agree, we should at least support operator and fuzziness.

Sounds good, I think I'd like to add them in a follow up

jimczi commented:
> Sounds good, I think I'd like to add them in a follow up

Sure as you like. Let's not forget it though ;)

andyb-elastic commented:
I'm adding those parameters here since it's not a big change

andyb-elastic commented:
I pushed changes that rename the query to `match_bool_prefix` and provide support for `operator` and `fuzziness` and its related parameters

andyb-elastic commented:
master - 6bba9fc83bc30d29c67e66e8e4c70f4dcc3b7494
7.x - 23395a9b9fd40f83a3fd5b8a5a7f3a05030a6743

telendt commented on modules/mapper-extras/src/test/resources/rest-api-spec/test/search-as-you-type/20_highlighting.yml:
```diff
@@ -0,0 +1,202 @@
+  - match: { hits.total: 1 }
+  - match: { hits.hits.0._source.a_field: "quick brown fox jump lazy dog" }
+  - match: { hits.hits.0._source.text_field: "quick brown fox jump lazy dog" }
+  - match: { hits.hits.0.highlight.a_field.0: "quick <em>brown</em> fox jump lazy dog" }
```
@andyb-elastic Is it unreasonable to expect `quick <em>brown</em> <em>fo</em>x jump lazy dog` highlight here?

telendt replied:
I know it's not nice to comment on already merged PR, that's why I opened this issue:
https://github.com/elastic/elasticsearch/issues/53744
